curvefs/src/client/rpcclient/cli2_cient.cpp

```diff
@@ -0,0 +1,150 @@
+/*
+ *  Copyright (c) 2021 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+
+/*
+ * Project: curve
+ * Created Date: Mon Sept 4 2021
+ * Author: lixiaocui
+ */
+
+#include "curvefs/src/client/rpcclient/cli2_client.h"
+
+namespace curvefs {
+namespace client {
+namespace rpcclient {
+
+void Cli2Closure::Run() {
+    bool success = false;
+    if (cntl.Failed()) {
+        LOG(WARNING) << "GetLeader failed from " << cntl.remote_side()
+                     << ", logicpool id = " << taskContext.poolID
+                     << ", copyset id = " << taskContext.copysetID
+                     << ", error = " << cntl.ErrorText();
+    } else {
+        LOG(INFO) << "GetLeader returned from " << cntl.remote_side()
+                  << ", logicpool id = " << taskContext.poolID
+                  << ", copyset id = " << taskContext.copysetID
+                  << ", leader = " << response.DebugString();
+        success = true;
+    }
+
+    excutor->NotifyRpcFinish(success);
+}
+
+#define RPCFunc [&](brpc::Channel * channel) -> int  // NOLINT
+
+bool Cli2ClientImpl::GetLeader(const LogicPoolID &poolID,
+                               const CopysetID &copysetID,
+                               const PeerInfoList &peerInfoList,
+                               int16_t currentLeaderIndex, PeerAddr *peerAddr,
+                               MetaserverID *metaserverID) {
+    int16_t index = -1;
+    peerAddr->Reset();
+
+    bool getLeaderOK = false;
+    // TODO(pri0@lixiaocui): optimization with back up request
+    for_each(peerInfoList.begin(), peerInfoList.end(),
+             [&](const CopysetPeerInfo<MetaserverID> &info) {
+                 ++index;
+
+                 if (index == currentLeaderIndex || getLeaderOK) {
+                     return;
+                 }
+
+                 std::string senderAddr(
+                     butil::endpoint2str(info.externalAddr.addr_).c_str());
+
+                 auto excutor = std::make_shared<TaskExecutor2>();
+                 Cli2TaskContext taskCtx(poolID, copysetID, senderAddr);
+                 Cli2Closure *done = new Cli2Closure(taskCtx, excutor);
```

Comment:
`done` maybe leaked, you should check it.

Reply:
fix